### PR TITLE
iOS,macOS: Add Obj-C cflags to all Obj-C targets

### PR DESCRIPTION
**Pull request by cbracken**
Ensure that all Objective-C code in the codebase is being built with the standard set of Flutter Objective-C compiler flags with ARC enabled.

Also bumps the cflags config up to the top of the first block within each target in which Objective-C sources appear, so that the location is consistent.

Migrates The following targets to ARC, which had been missed in previous passes since they didn't declare the standard Flutter Obj-C[++] cflags:
* `//flutter/fml:fml_unittests`
* `//flutter/impeller/golden_tests:metal_screenshot`
* `//flutter/impeller/playground:playground`
* `//flutter/impeller/backend/metal:metal`
* `//flutter/impeller/backend/metal:metal_unittests`
* `//shell/gpu:gpu_surface_metal_unittests`
* `//flutter/shell/platform/embedder:embedder_unittests`

This patch includes no semantic changes.

Issue: https://github.com/flutter/flutter/issues/137801

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I signed the [CLA].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: fml/BUILD.gn [hunk 1]
```diff
@@ -365,7 +362,10 @@ if (enable_unittests) {
       "time/time_unittest.cc",
     ]
 
-    if (is_mac) {
+    if (is_mac || is_ios) {
```
Comment (cbracken) [inline fml/BUILD.gn:365]: Fix: making consistent with the associated conditional in the target above that this one is testing.

File: shell/platform/embedder/tests/embedder_metal_unittests.mm [hunk 2]
```diff
@@ -65,7 +65,7 @@
                                               SkISize texture_size,
                                               void* texture) {
   GrMtlTextureInfo info;
-  info.fTexture.reset([(id<MTLTexture>)texture retain]);
+  info.fTexture.retain(texture);
```
Comment (cbracken) [inline shell/platform/embedder/tests/embedder_metal_unittests.mm:68]: Found the non-arc code!

File: impeller/BUILD.gn [hunk 3]
```diff
@@ -80,6 +79,10 @@ impeller_component("impeller_unittests") {
     ]
   }
 
+  if (impeller_enable_metal) {
+    deps += [ "//flutter/impeller/renderer/backend/metal:metal_unittests" ]
+  }
```
Comment (cbracken) [inline impeller/BUILD.gn:84]: No need to build these guys on windows, linux, or custom embedders for platforms without metal enabled.

**Review by hellohuanlin — APPROVED** [review]

File: impeller/BUILD.gn [hunk 4]
```diff
@@ -60,7 +60,6 @@ impeller_component("impeller_unittests") {
     "core:allocator_unittests",
     "display_list:skia_conversions_unittests",
     "geometry:geometry_unittests",
-    "renderer/backend/metal:metal_unittests",
```
Comment (jmagman) [inline impeller/BUILD.gn:63]: How was this passing if impeller_enable_metal was false?

**Review by jmagman — APPROVED** [review]

File: impeller/renderer/backend/metal/BUILD.gn [hunk 5]
```diff
@@ -5,6 +5,9 @@
 import("../../../tools/impeller.gni")
 
 impeller_component("metal") {
+  cflags_objc = flutter_cflags_objc_arc
+  cflags_objcc = flutter_cflags_objcc_arc
```
Comment (stuartmorgan) [inline impeller/renderer/backend/metal/BUILD.gn:9]: I skimmed some of the files in the list below, and a bunch have Obj-C ivars ([example](https://github.com/flutter/engine/blob/382fd1ea143afd57c1191500d75ac23984e939cf/impeller/renderer/backend/metal/blit_pass_mtl.h#L22-L24)). Is implicitly changing those from weak to strong intentional?

Comment (cbracken) [reply impeller/renderer/backend/metal/BUILD.gn:9]: Re-auditing, you're right there are some in here that should be weak. There were a lot of cases where they should be strong -- i.e. pre-ARC code was allocating but never releasing objects, which I had been using as a shortcut to find the strong objects; looks like that shortcut doesn't apply well here. I'm going to break this patch up.

Comment (cbracken) [reply impeller/renderer/backend/metal/BUILD.gn]: Reverted -- only change now is standardisation of flags and ARC-ificiation of simple unit tests that hadn't been migrated. Impeller to follow.

**Pull request merged**